### DISCUSSION TITLE
Dual stack (IPv4/IPv6) support

### DISCUSSION THREAD

**Discussion by oktavlachs** [Kludex/uvicorn#1529] [Ideas] [answered]
Hello everyone and thanks for your work on this great project :smiley_cat: 

I find it especially helpful in combination with FastAPI, which is my use case.

What I cannot use is the dual stack mode (combined IPv4/IPv6) on the socket used by Uvicorn, which would come in handy in environments in which both IP versions are supported. In this case, I would assume to use the IPv6 host address provided by the user and then add the dual stack mode on top.

From my point of view this would be a pretty straight forward implementation by extending the config method `bind_socket` [HERE](https://github.com/encode/uvicorn/blob/6286ecc398f8e3c368ba945c2f481cfba36ca0f8/uvicorn/config.py#L549) when starting a server. You would need to add an OS check and then configure the socket like this:

``` Python
if socket.has_dualstack_ipv6():
    sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 0)
```
Is this something you would support in the product?
If so, would a PR be welcome?

Cheers and have a nice evening,
Oktavlachs

**Comment by euri10** [comment]
nothing against it, question is how is it testable ?

**Reply by oktavlachs** [reply]
Integration test could look similar to [the existing basic one](https://github.com/encode/uvicorn/blob/39621c20b44187f9d31d5348d34aeab400436170/tests/test_main.py#L62).
But you use the httpx client for both, an IPv6 GET request and after that an IPv4 GET request.

What do you think?

**Reply by euri10** [reply]
ok I was looking at socket.py and `create_server` which we are using, it seems to me it detects if y/n it has dualstack enabled.
So now I'm confused as to what you're trying to achieve you cant ?
Or I'm mixing things, entirely possible, care to explain @oktavlachs ?

**Reply by oktavlachs** [reply]
Thanks for your update. Actually, I solved my issue by handing over my own configured socket into the `Server.serve()` method. I wasn't aware that this is possible, but found out by the latest documentation update which will be merged into the next release (programmatic use of uvicorn).

However, I still think you should also be able to start dual stack from the command line and therefore the `run` method in the `main.py` script should be handed over with a particularly configured socket for dual stack. Does that make sense to you?

In my opinion this wouldn't need additional tests, since you are using the existing standard library function, which from my point of view doesn't mess up other logic. Am I correct here?

**Reply by euri10** [reply]
> However, I still think you should also be able to start dual stack from the command line and therefore the `run` method in the `main.py` script should be handed over with a particularly configured socket for dual stack. Does that make sense to you?

yes it does, that said I don't think it is necessary since as I said the create_server handles that for you. If this is not the case then could you show either a PR or what you have in mind, will be clearer for me >_)

**Reply by oktavlachs** [reply]
Please find a simple script to show the difference between the `create_server` function of the `socket.py` module you are referring to and the asyncio `create_server` function used in uvicorn [HERE](https://github.com/encode/uvicorn/blob/ad7b00a6e092cebf5d32358d2d1a94bcb9931457/uvicorn/server.py#L149).

```Python
import asyncio
from asyncio import BaseProtocol
import socket


def run_socket_server():
    socket.create_server(('::', 8088), family=socket.AF_INET6, dualstack_ipv6=True)
    print("Socket create_server OK")


class RandomProtocol(BaseProtocol):
    pass


async def run_async_server():
    await asyncio.get_running_loop().create_server(
        RandomProtocol,
        host='::',
        port=8089,
        # Comment/uncomment this one
        dualstack_ipv6=True
    )
    print("Asyncio create_server OK")


# Run socket server
run_socket_server()

# Run async server
asyncio.run(run_async_server())

```

If you run the code above as is, it will throw a TypeError:

```
Socket create_server OK
Traceback (most recent call last):
  File "/main.py", line 30, in <module>
    asyncio.run(run_async_server())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/main.py", line 16, in run_async_server
    await asyncio.get_running_loop().create_server(
TypeError: BaseEventLoop.create_server() got an unexpected keyword argument 'dualstack_ipv6'
```

So I would say, a pre-configured socket for the asyncio `create_server` function should do the job, because the `dualstack_ipv6` parameter is not available.

So I would proceed with a PR and give an update here?

**Reply by euri10** [reply]
mmm my bad I was indeed on the wrong create_server ....I mean not the one we're using

> So I would proceed with a PR and give an update here?

yes please

**Reply by Kludex** [reply]
As an alternative, we can just document how to achieve this via python code.

We aim to match CLI way with python code way instinctively, but not all the options are available through CLI right now.

We can also not add another argument, but actually set as none when the argument is passed with empty value. 🤔

**Reply by Kludex** [reply]
A message was deleted here...

**Reply by oktavlachs** [reply]
@euri10 
Thanks for your support. Actually, I found that dual stack mode can be achieved already by setting the host parameter in the uvicorn Config object to `None` (needs minor code changes).
So like this:

```Python
ipv4_config = Config(app=_fastapi_server, host=None, port=8080)

return Server(ipv4_config)
```

Uvicorn will then hand over None as a value to the loop.create_server() function, which leads to it listening on all interfaces including both IPv4 and IPv6 in dual stack mode. This is confusing, since we both would have expected the function to have a parameter for this, as the socket.create_server() function.

From the official Python [documentation](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_server):

`If host is an empty string or None, all interfaces are assumed and a list of multiple sockets will be returned (most likely one for IPv4 and another one for IPv6).`

However, this works fine if you use uvicorn directly in your code, but the CLI currently cannot be used for this, since it always has a host parameter input. So my suggestion is to introduce a new --dualstack parameter, which then overrides whatever was provided as a host and listens on all interfaces.

**Reply by oktavlachs** [reply]
@Kludex That was me by accident

> A message was deleted here...

**Reply by oktavlachs** [reply]
> As an alternative, we can just document how to achieve this via python code.
> 
> We aim to match CLI way with python code way instinctively, but not all the options are available through CLI right now.
> 
> We can also not add another argument, but actually set as none when the argument is passed with empty value. thinking

The last one would be the best of course, but I would consider it a breaking change, since it changes the way CLI behaves. It is conservative by default (only localhost), which is good security wise.

**Reply by oktavlachs** [reply]
So I would vote for another parameter, which enables users to use dual stack but doesn't break existing implementations

**Reply by oktavlachs** [reply]
E.g.

`--dualstack If enabled, the --host parameter is ignored and uvicorn will listen on all interfaces for both, IPv4 and IPv6.`

**Reply by oktavlachs** [reply]
> We can also not add another argument, but actually set as none when the argument is passed with empty value. thinking

I am sorry, you are right. If there is currently no implementation for an empty string, we could also use this. Before causing more confusion, let's talk after I did a PR

**Reply by Kludex** [reply]
Are you still interested in this? @oktavlachs

**Reply by oktavlachs** [reply]
@Kludex Yes I am. Let me come back to this by Sunday evening. 👍

**Reply by 17336** [reply]
> @euri10 Thanks for your support. Actually, I found that dual stack mode can be achieved already by setting the host parameter in the uvicorn Config object to `None` (needs minor code changes). So like this:
> 
> ```python
> ipv4_config = Config(app=_fastapi_server, host=None, port=8080)
> 
> return Server(ipv4_config)
> ```
> 
> Uvicorn will then hand over None as a value to the loop.create_server() function, which leads to it listening on all interfaces including both IPv4 and IPv6 in dual stack mode. This is confusing, since we both would have expected the function to have a parameter for this, as the socket.create_server() function.
> 
> From the official Python [documentation](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_server):
> 
> `If host is an empty string or None, all interfaces are assumed and a list of multiple sockets will be returned (most likely one for IPv4 and another one for IPv6).`
> 
> However, this works fine if you use uvicorn directly in your code, but the CLI currently cannot be used for this, since it always has a host parameter input. So my suggestion is to introduce a new --dualstack parameter, which then overrides whatever was provided as a host and listens on all interfaces.

it works when app.run(debug=False). But if set debug=True, when host is empty it only listen ipv4 and host is None will raise an error.

**Reply by Kludex** [reply]
The latest version of Uvicorn doesn't even have the `debug` parameter. :thinking:

**Reply by rogue73** [reply]
Hi, how is the status about support of dual stack?

**Reply by pkit** [reply]
I see that it doesn't work, too bad. Async "support" in python is at its "best" again. Lol

**Reply by pythonwood** [reply]
2024-02i test uvicorn  0.20  in ubuntu and macos , `uvicorn --host '::' --port 8000` support  both ipv6/ipv4
ss check output `TCP *:8000 (LISTEN)`
curl test by `curl 127.0.0.1:8000  '[::1]:8000'` 
every things ok.

**Comment by bmx666** [comment]
Hi, any updates?

**Reply by QwertyJack** [reply]
+1

**Comment by bulletmark** [comment]
I've been going down this rabbit hole but it seems to me that dual stack IPv4/IPv6 support is disabled in Python asyncio as per [these lines of code](https://github.com/python/cpython/blob/main/Lib/asyncio/base_events.py#L1605) so I don't think anything can be done about this in uvicorn.

**Comment by kamyabnazari** [comment]
Hello everyone, so is there an alternative that I can us to deploy and make the dual stack possible with my fast api backend or do I have to switch to another backend framework completely, because right now in my case node js fetch prefers ipv6 and I am not mapping anything on my fast api backend to ipv6 and as such it fails because of "::1" error. So if this is not possible right is there an alternative you guys know about? 

Thank you.

**Reply by rogue73** [reply]
@kamyabnazari Hi, in FastAPI projects where I need dualstack support, I currently choosed to use hypercorn.

**Reply by pkit** [reply]
@rogue73 I think you need to add that to use hypercorn you MUST bind twice.
I.e. `hypercorn --bind '[::1]:8888' --bind '127.0.0.1:8888' ...` otherwise it does not bind dualstack too.

**Reply by pythonwood** [reply]
test in linux hypercorn 0.13.2
but --bind [::]:8888 just equal to --bind '0.0.0.0:8888'， no ipv6

update: may be a linux os problem but not a macos problem.

thans for [ test ](https://github.com/benoitc/gunicorn/issues/1138#issuecomment-158358666)

**Comment by Kludex** [comment, marked as answer]
https://github.com/encode/uvicorn/discussions/1529#discussioncomment-3061823

**Comment by pythonwood** [comment]
Testing uvicorn  0.20  in ubuntu and macos , `uvicorn --host '::' --port 8000` support  both ipv6/ipv4
ss check output `TCP *:8000 (LISTEN)`
curl test by `curl 127.0.0.1:8000  '[::1]:8000'` 
every thing is ok.

**Reply by mcanevet** [reply]
@pythonwood somehow it does not work for me. I tested with `0.27.1` and it works on IPv6 but not IPv4 with `--host '::'`

**Reply by pythonwood** [reply]
@mcanevet it seems need another condition **reload**
In fact, I run direct _**python3 main.py**_.
listening ipv4 and ipv6 *:8000 only comes when **reload=True** in my main.py. 
...
`uvicorn.run(app, host='::', port=8000, reload=True)`
...
when reload=False, it only listens addr [::]:8000.

**Reply by pythonwood** [reply]
because `python3 -m http.server --bind ::` can listen on *:8000 (both ip4+6)
hope uvicorn can do same things

**Reply by philip-couling-acs** [reply]
I suspect what you may be encountering is a particular behavior for a specific C library and may not be completely ubiquitous.

There's some mention of it on [this man page](https://man7.org/linux/man-pages/man7/ipv6.7.html#:~:text=IPv4%20connections%20can%20be%20handled%20with%20the%20v6%20API%20by%20using%20the%0A%20%20%20%20%20%20%20v4%2Dmapped%2Don%2Dv6%20address%20type%3B%20thus%20a%20program%20needs%20to%20support%20only%0A%20%20%20%20%20%20%20this%20API%20type%20to%20support%20both%20protocols.%20%20This%20is%20handled%0A%20%20%20%20%20%20%20transparently%20by%20the%20address%20handling%20functions%20in%20the%20C%20library)

AFAIK this can't be relied on for all OS and all C implementations and is more generally discouraged, although I'm struggling to lay my hands on the recommendation for this.

**Comment by Strangerxxx** [comment]
I've just tried to fix this in the fork
Take a look on PR: https://github.com/encode/uvicorn/pull/2266

**Comment by hanxi** [comment]
Both IPv4 and IPv6 are supported, although this is not documented. The following code snippet demonstrates this functionality:

```python
uvicorn.run(app, host=["::", "0.0.0.0"], port=8080)
```

From <https://github.com/hanxi/xiaomusic/blob/e2e74f9459530a2bcdc8cfaf9a294e7e9e677c8b/xiaomusic/cli.py#L80>

**Reply by amn** [reply]
Is it? Looking at https://github.com/encode/uvicorn/blob/5bf788f0eb0fc771f5c4eed8f282c7ec256565d2/uvicorn/main.py#L363, the typing hint for `host` is `str`, which to me suggests list values cannot be passed with `host` to the procedure. Looking also at [the `bind_socket` procedure](https://github.com/encode/uvicorn/blob/5bf788f0eb0fc771f5c4eed8f282c7ec256565d2/uvicorn/config.py#L479-L525) I don't see anything that would suggest there's any looping over values specified with `host`. This is all from `master`, mind you.

Which branch of Uvicorn are you using?

**Reply by hanxi** [reply]
@amn
Uvicorn version is 0.30.1 , The log is :

```
Uvicorn running on http://['::', '0.0.0.0']:3333 (Press CTRL+C to quit)
```

netstat output

```
sudo netstat -tunlp|grep 3333
tcp        0      0 0.0.0.0:3333            0.0.0.0:*               LISTEN      522371/python
tcp6       0      0 :::3333                 :::*                    LISTEN      522371/python
```

**Reply by amn** [reply]
Right, thank you -- found now that the sockets are bound by https://docs.python.org/3.12/library/asyncio-eventloop.html#asyncio.loop.create_server and not by the aforementioned `bind_socket` procedure. The former supports list for `host`:

> If host is a sequence of strings, the TCP server is bound to all network interfaces specified by the sequence.

**Reply by uwu-420** [reply]
Can confirm this works!
@Kludex would there be something speaking against updating the type hint for `host` then from `str` to `str | list[str]` ?

**Comment by mcanevet** [comment]
It would still be great to be able to set this through command line because one do not always have access to the code...

**Reply by ToroNZ** [reply]
This works on Uvicorn 0.30.6

`CMD ["uvicorn", "main:app", "--host", "", "--port", "80"]`

```bash
INFO:     Uvicorn running on http://:80 (Press CTRL+C to quit)              # <-- Binds to all interfaces
INFO:     fd34:acc1:abfd::1:48534 - "GET /openapi.json HTTP/1.1" 200 OK     # <-- Reachable on [::1]
INFO:     172.17.0.1:40830 - "GET /openapi.json HTTP/1.1" 200 OK            # <-- Reachable on 127.0.0.1
```

**Reply by MichalMMac** [reply]
This works nicely in most environments but Google Cloud Run environment really does not like the empty string `, "",` in `CMD` array for some reason.

**Update:** The workaround is to create a simple shell script and set it as entrypoint. There is not issue with using the empty string inside of the script.

**Reply by b-a-t** [reply]
That's the only suggestion that worked for me as of `uvicorn[standard]==0.30.6`

**Comment by RemiPaulmier** [comment]
According to python's doc (https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_server), the host parameter in create_server is either a string, a sequence of strings, or None/empty string. 
Thx to @ToroNZ we know empty string works, but it is both not very explicit and passing an empty string can be challenging in some environments.

A simple change like this in `server.py` does the job: 
```
server = await loop.create_server(
    create_protocol,
    host=config.host.split(","),
    [...]
```

and allows more explicit calls like:

`CMD ["uvicorn", "main:app", "--host", "::,0.0.0.0", "--port", "80"]`

This produces the following log:
```
[...]
INFO:     Uvicorn running on http://[::,0.0.0.0]:80 (Press CTRL+C to quit)
[...]
```

netstat output:
```
tcp       0      0 0.0.0.0:80            0.0.0.0:*               LISTEN      149755/python
tcp6      0      0 :::80                 :::*                    LISTEN      149755/python
```
